File: server/system_files/src/utils/parser.py
import sys
import json
import re
from urllib.parse import urlparse, quote

def sanitize_url(url):
    """Prevents n8n double-encoding chokes."""
    if not url or not isinstance(url, str):
        return None
    if 'vertexaisearch' in url or 'grounding-api-redirect' in url:
        return None
    try:
        parsed = urlparse(url.strip())
        return f"{parsed.scheme}://{parsed.netloc}{quote(parsed.path)}{parsed.params}{parsed.query}{parsed.fragment}"
    except:
        return url

def strip_citations(val):
    """Strips [1], [2, 3] etc."""
    if not isinstance(val, str):
        return val
    return re.sub(r'\[\d+(?:,\s*\d+)*\]', '', val).strip()

def extract_friction(data):
    """Heuristic search for forensic friction terms."""
    enriched = data.get('Enriched_Lead', {})
    divers = enriched.get('The_Divers', data.get('The_Divers', {}))
    
    candidates = [
        enriched.get('Forensic_Friction_Type'),
        enriched.get('friction_type'),
        divers.get('friction_notes'),
        enriched.get('friction_notes'),
        data.get('friction_type')
    ]

    forensic_terms = ['API Stutter', 'Scale Friction', 'Manual Data Debt', 'Displacement Signal']
    
    for val in candidates:
        if not isinstance(val, str): continue
        for term in forensic_terms:
            if term.lower() in val.lower():
                return term
    return None

def normalize_batch(payload):
    """Normalizes n8n payloads for the server."""
    raw_leads = []
    if isinstance(payload, dict):
        raw_leads = payload.get('leads') or [payload]
    elif isinstance(payload, list):
        raw_leads = payload

    processed = []
    for item in raw_leads:
        data = item.get('json', item) if isinstance(item, dict) else item
        
        sid = data.get('session_id') or data.get('Session_ID')
        cname = data.get('company_name') or data.get('Company_Name')
        
        if sid and cname:
            processed.append({
                "session_id": sid,
                "company_name": cname,
                "email": data.get('email'),
                "friction_type": extract_friction(data)
            })
    return processed

if __name__ == "__main__":
    try:
        input_data = json.load(sys.stdin)
        result = normalize_batch(input_data)
        print(json.dumps(result))
    except Exception as e:
        print(json.dumps({"error": str(e)}), file=sys.stderr)
        sys.exit(1)

File: server/system_files/src/utils/parser.js
/**
 * Sanitizes URLs to prevent n8n "double-encoding" chokes.
 * Uses encodeURIComponent on the path while preserving the protocol and host.
 */
export function sanitizeUrl(url) {
  if (!url || typeof url !== 'string') return null;
  if (url.includes('vertexaisearch') || url.includes('grounding-api-redirect')) return null;
  
  try {
    const parsed = new URL(url.trim());
    // n8n compliance: pre-encoding the path to avoid double-encoding in HTTP nodes
    return `${parsed.protocol}//${parsed.host}${encodeURI(parsed.pathname)}${parsed.search}${parsed.hash}`;
  } catch (e) {
    return url; // Fallback to raw if not a valid URL
  }
}

/**
 * Strips bracketed citations [1], [2, 3] from strings.
 */
export function stripCitations(val) {
  if (typeof val !== 'string') return val;
  return val.replace(/\[\d+(?:,\s*\d+)*\]/g, '').trim();
}

/**
 * Extracts JSON from a potentially markdown-wrapped model response.
 */
export function extractJson(raw) {
  const cleaned = raw.trim();
  try {
    // Attempt 1: Direct parse
    return JSON.parse(cleaned);
  } catch (e) {
    // Attempt 2: Regex-based extraction (Outer most braces or brackets)
    const match = cleaned.match(/[\{\[]([\s\S]*)[\}\]]/);
    if (match) {
      try {
        return JSON.parse(match[0]);
      } catch (innerError) {
        throw new Error(`Failed to parse extracted JSON: ${innerError.message}`);
      }
    }
    throw new Error('No valid JSON structure found in response');
  }
}

/**
 * Extracts specific forensic friction types from the payload.
 */
export function extractFrictionType(payload) {
  const lead = payload?.Enriched_Lead || payload || {};
  const divers = lead.The_Divers || payload?.The_Divers || {};
  
  const candidates = [
    lead.Forensic_Friction_Type,
    lead.friction_type,
    divers.friction_notes,
    lead.friction_notes
  ];

  const forensicTerms = [
    'API Stutter', 
    'Scale Friction', 
    'Manual Data Debt', 
    'Displacement Signal'
  ];

  const serviceIntents = ['GTM', 'Accounting'];

  for (const val of candidates) {
    if (typeof val !== 'string') continue;
    const found = forensicTerms.find(term => val.toLowerCase().includes(term.toLowerCase()));
    if (found) return found;
    if (serviceIntents.some(intent => val.trim().toUpperCase() === intent)) continue;
    if (forensicTerms.includes(val.trim())) return val.trim();
  }

  return null;
}

/**
 * Extracts funding signals from various nested properties.
 */
export function extractFundingSignal(payload) {
  if (payload?.Enriched_Lead?.funding_signal) return payload.Enriched_Lead.funding_signal;
  if (payload?.funding_signal) return payload.funding_signal;
  if (payload?.Nemo_Enrich_Audit?.funding_signal) return payload.Nemo_Enrich_Audit.funding_signal;

  const notes =
    payload?.The_Divers?.health_audit_notes ||
    payload?.Enriched_Lead?.The_Divers?.health_audit_notes ||
    '';
    
  if (typeof notes === 'string' && notes) {
    const m = notes.match(/\$[\d.,]+\s*[MBK]?(?:\s*million|\s*billion)?|Series\s+[A-E]|Seed\s+(?:round)?|Pre-?[Ss]eed/i);
    if (m) return m[0];
  }
  return null;
}

/**
 * Normalizes Nemo payload for outreach synthesis.
 */
export function normalizeNemoPayload(payload) {
  const email = 
    payload?.Contact_Recon?.email || 
    payload?.Contact_Recon?.email_pattern || 
    payload?.Contact_Recon?.email_pattern_guess || 
    payload?.contact_recon?.email || 
    payload?.Enriched_Lead?.Contact_Recon?.email || 
    payload?.email || 
    null;

  const contactRaw = 
    payload?.Contact_Recon || 
    payload?.contact_recon || 
    payload?.Enriched_Lead?.Contact_Recon || 
    null;

  return {
    email,
    contact_recon: contactRaw || null,
    friction_type: extractFrictionType(payload)
  };
}
